File: lib/to_mdx/parsers/open_document/text/text_box.rb
# frozen_string_literal: true

require_relative "../base"
require_relative "../../../detectors/code_detector"

module ToMdx
  module Parsers
    module Text
      class TextBox < Base
        include ToMdx::Detectors::CodeDetector

        attr_reader :text_style_name

        def initialize(element = nil, context = {})
          super(element, context)
          @text_style_name = context[:styles].use(element&.attributes&.[]("draw:text-style-name"))
        end

        def to_mdx
          code_snippet = maybe_code_snippet_mdx
          return code_snippet if code_snippet

          content = contentful_children.map(&:to_mdx).join(" ")
          if @text_style_name.empty?
            content
          else
            "<span className=\"#{@text_style_name}\">#{content}</span>"
          end
        end

        def to_s
          contentful_children.map(&:to_s).join
        end

        def with_context(context)
          context.merge(
            parent_is_code_snippet: is_using_monospace_font?
          )
        end

        register_for "draw:text-box"
      end
    end
  end
end

File: src/parsers/open-document/text/text_box.ts
import { CodeDetector } from "../../../detectors/code-detector.ts";
import type { Style } from "../../../styles.ts";
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";

export class TextBox extends BaseElement {
  private textStyleName: Style;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    this.textStyleName = context.styles.use(this.attr("draw:text-style-name"));
  }

  toMdx(): string {
    const codeSnippet = this.codeDetector.maybeCodeSnippet();
    if (codeSnippet) {
      return codeSnippet;
    }

    const content = this.contentfulChildren()
      .map((child) => child.toMdx())
      .join(" ");

    // Text boxes are always preformatted text in ODP
    const className = this.textStyleName.empty() ? "" : ` className="${this.textStyleName}"`;
    return `<pre${className}>${content}</pre>`;
  }

  toString(): string {
    return this.contentfulChildren()
      .map((child) => child.toString())
      .join("");
  }

  private get codeDetector(): CodeDetector {
    return new CodeDetector(this, () => this.toString());
  }

  static {
    BaseElement.registerFor.call(this, "draw:text-box");
  }
}
